File: anidesk_py/src/anidesk/services/season.py
from __future__ import annotations

import re
import unicodedata
from datetime import date, datetime

from anidesk.domain.models import Season

SEASON_MONTHS: dict[Season, tuple[int, ...]] = {
    Season.WINTER: (1, 2, 3),
    Season.SPRING: (4, 5, 6),
    Season.SUMMER: (7, 8, 9),
    Season.FALL: (10, 11, 12),
}

SEASON_LABELS: dict[Season, str] = {
    Season.WINTER: "冬季",
    Season.SPRING: "春季",
    Season.SUMMER: "夏季",
    Season.FALL: "秋季",
}


def season_for_month(month: int) -> Season:
    if not 1 <= month <= 12:
        raise ValueError("month must be between 1 and 12")
    if month <= 3:
        return Season.WINTER
    if month <= 6:
        return Season.SPRING
    if month <= 9:
        return Season.SUMMER
    return Season.FALL


def current_season(value: date | datetime | None = None) -> tuple[int, Season]:
    current = value or datetime.now()
    return current.year, season_for_month(current.month)


def normalize_title(value: str) -> str:
    normalized = unicodedata.normalize("NFKC", value).casefold()
    return re.sub(r"[\W_]+", "", normalized, flags=re.UNICODE)


def iso_date(year: int | None, month: int | None, day: int | None) -> str:
    if not year or not month or not day:
        return ""
    return f"{year:04d}-{month:02d}-{day:02d}"

File: src/services/season.ts
import type { Season } from "../types";

const SEASON_MONTHS: Record<Season, number[]> = {
  WINTER: [1, 2, 3],
  SPRING: [4, 5, 6],
  SUMMER: [7, 8, 9],
  FALL: [10, 11, 12],
};

export const SEASON_LABELS: Record<Season, string> = {
  WINTER: "冬季",
  SPRING: "春季",
  SUMMER: "夏季",
  FALL: "秋季",
};

export function getSeasonForMonth(month: number): Season {
  if (month <= 3) return "WINTER";
  if (month <= 6) return "SPRING";
  if (month <= 9) return "SUMMER";
  return "FALL";
}

export function currentSeason(date = new Date()): { year: number; season: Season } {
  return { year: date.getFullYear(), season: getSeasonForMonth(date.getMonth() + 1) };
}

export function monthsForSeason(season: Season): number[] {
  return [...SEASON_MONTHS[season]];
}

export function normalizeTitle(value: string): string {
  return value
    .normalize("NFKC")
    .toLocaleLowerCase()
    .replace(/[\s\p{P}\p{S}]/gu, "");
}

export function isoDate(parts: { year?: number | null; month?: number | null; day?: number | null }): string {
  if (!parts.year || !parts.month || !parts.day) return "";
  return `${parts.year.toString().padStart(4, "0")}-${parts.month.toString().padStart(2, "0")}-${parts.day
    .toString()
    .padStart(2, "0")}`;
}
